feat(remix-dev/vite, remix-react, remix-server-runtime): support custom basename

hi-ogawa opened the pull request:
Closes: https://github.com/remix-run/remix/issues/8052
References: https://github.com/remix-run/remix/discussions/2891, https://github.com/remix-run/remix/pull/8077#issuecomment-1823469405, https://github.com/remix-run/remix/pull/5236

This PR further extends https://github.com/remix-run/remix/pull/8141 to integrate react-router's `basename` option so that route definition and navigation urls don't require to be manually prefixed with custom basename.
Maybe I should've discussed with Remix team beforehand, but I thought knowing the amount of change required for this feature can be an important factor, so here it is.
Please let me know if there is a concern about the introduction of this feature (and/or a concern about this PR's approach).
Thanks!

## summary of the changes

- Add `RemixVitePluginOptions.basename`
- Add `ServerBuild.basename` so that server-runtime can use it for `createStaticHandler(..., { basename })` and also for a few logic outside of react-router (e.g. "X-Remix-Redirect").
- Add `__remixContext.basename` so that client can use it for `createBrowserRouter(..., { basename })`.
- Replace `matchRoutes(routes, url)` with `matchRoutes(routes, url, basename)`.

Also the changes from https://github.com/remix-run/remix/pull/8141 are:

- Replace `req.url` with `req.originalUrl` since react-router's `StaticHandler` assumes request url includes basename.
- Add `publicPath` prefix to all client-visible url for vite dev.

## questions

- I noticed `v7_prependBasename` is introduced in [react-router#10336](https://github.com/remix-run/react-router/pull/10336). As far as my manual test goes (e.g. relative link, useNavigate, etc...), it seems to be working without this flag, but I could be missing something.

- For now, I'm only testing with Vite plugin. It might be possible to support non-Vite Remix, but I haven't attempted that yet. Please let me know if it's desired to add support for non-Vite use cases as well.

- Interaction between `basename` and `publicPath` options might be tricky. I was testing `basename != publicPath` case, but during dev, Vite server forces SSR request to come under the path of `base` (`= publicPath`) option, so it's required to have `basename.startsWith(publicPath)`. On the other hand, I believe there's no restriction for "build + start" as long as the server is configured to handle the routes properly. So, I added this exact validation in `resolvePluginConfig`, but probably we can explore other approaches as well (for example, if `publicPath` is undefined but `basename` is given, then remix automatically sets `publicPath = basename` assuming that's the most common usage).

## testing strategy

I setup a demo project in https://github.com/hi-ogawa/remix-vite-custom-base/pull/1 and wrote a list of features I'm testing manually.
For playwright test here, I only added very simple one as a reference. If the feature is agreed upon, then I plan to expand the test cases.

jrestall commented:
This is really awesome Hiroshi, as you mentioned and from my experience, it's definitely not super easy to host remix sites under a subfolder as the base url has to be manually added to every Link and redirect along with custom route generation. This will make it much easier thanks!

cevr commented:
one question - in loaders/actions a redirect could point to something outside of the basepath. Would this make it impossible to do so?

hi-ogawa commented:
> in loaders/actions a redirect could point to something outside of the basepath. Would this make it impossible to do so?

@cevr Thank for pointing out that use case. Remix/react-router has [`redirectDocument`](https://remix.run/docs/en/main/utils/redirectDocument) to force browser reload, so probably you can use this for such redirection. I added an example in the demo repository https://github.com/hi-ogawa/remix-vite-custom-base/pull/1/commits/53960850f619ef4668c7bca78bfdd2432a6b8b89

cevr commented:
@hi-ogawa do you know of any way to package this up so that it can be referenced on npm? This is the only thing blocking my work from migrating to remix 😩

kibertoad commented:
@markdalgleish Other than unsolved conflicts, are there any reasons why this PR wouldn't be accepted?

kiliman commented:
BTW: `remix-flat-routes` has a custom basename option. It will automatically append the basename to all your top-level routes.

https://github.com/kiliman/remix-flat-routes#api

cevr commented:
@kiliman wouldnt say its the same, as this effectively adds the same support that RR's basename does, making the base path invisible to the rest of the tree to both server and client

jrestall commented:
> BTW: `remix-flat-routes` has a custom basename option. It will automatically append the basename to all your top-level routes.
> 
> https://github.com/kiliman/remix-flat-routes#api

We are using this for basename support and it's great but there's still a ton of other changes required due to the router itself not having basename set. All redirects and Link usages need to have the basename manually added. The basename needs to be passed around the server and client as env variable. I'm excited for this PR since it should simplify subfolder hosting to a single config string.

kiliman commented:
True. Typically if I want to use basename to mount routes to a non-root folder, I ensure that my app always uses relative routes, so it doesn't matter what the "root" is.

It would be nice to have native support for this though.

brophdawg11 commented on packages/remix-dev/vite/plugin.ts:
```diff
@@ -488,14 +483,19 @@ export const remixVitePlugin: RemixVitePlugin = (remixUserConfig = {}) => {
-        "'publicPath' must be a prefix of 'basename' for default Vite dev server."
-      );
+      if (!basename.startsWith(publicPath)) {
+        console.warn(
```
I changed this from an `invariant` to a warning since the invariant ends up including a message like `"This is a bug in Remix"` in it

brophdawg11 commented on packages/remix-express/server.ts:
```diff
@@ -94,8 +94,7 @@ export function createRemixRequest(
   let resolvedHost = `${req.hostname}${port ? `:${port}` : ""}`;
-  // TODO: could it break existing hacky custom base use cases?
-  let url = new URL(`${req.protocol}://${resolvedHost}${req.originalUrl}`);
+  let url = new URL(`${req.protocol}://${resolvedHost}${req.url}`);
```
I didn't seem to need this in my testing of a remix vite express app using a basename - was this intended for apps using custom middlewares that mess with `req.url`?

hi-ogawa replied:
This probably depends on the way express routers/middlewares are structured.

In my example setup https://github.com/hi-ogawa/remix-vite-custom-base/pull/1, I used a separate `express.Router` to handle remix related things completely separately, which looks like this:

```ts
// [ROOT ROUTER] handle "/..."
const rootApp = express()

// things not related to remix goes here...
rootApp.get("/", ...) // 
rootApp.get("/not-remix", ...) 

// [SUB ROUTER] handle "/mybase/..." as a separate route for remix
const remixApp = express.Router()
rootApp.use("/mybase", remixApp)

// usual remix custom server setup goes here...
remixApp.use(viteDevServer.middlewares)
remixApp.all("*", createRequestHandler({...}));

// finally listen at the root
rootApp.listen()
```

In this case, `url = strip(originalUrl, "/mybase")` is probably done by express when nesting router, so this would be "custom middleware that mess with `req.url`" like you suggested.

I'm trying to find how to avoid such stripping by middlewares, but I cannot figure it out. For example, replacing above `remixApp` with `rootApp` to avoid intermediate router doesn't seem to work, which is probably because of `viteDevServer.middlewares` also strips basename https://github.com/vitest-dev/vite/blob/06494443194763d6dd125740961cec82dd9a29d6/packages/vite/src/node/server/middlewares/base.ts#L21-L26.

Could you share how your express server is setup?

brophdawg11 replied:
ah ok yeah I had just a very simple setup with static base paths and no middlewares/sub-routers: 

```js
const app = express();

app.use("/base/", viteDevServer?.middlewares || express.static("build/client"));

app.all(
  "/base/*",
  createRequestHandler({
    build: viteDevServer
      ? () => viteDevServer.ssrLoadModule("virtual:remix/server-build")
      : await import("./build/server/index.js"),
  })
);
```

I like that workaround you found to re-assign so I agree that it's safest to stick with `req.url` now and if folks are messing with the URL ahead of time they can reset it 👍

brophdawg11 commented on packages/remix-react/browser.tsx:
```diff
@@ -24,7 +24,7 @@ import {
   var __remixContext: {
     url: string;
-    basename?: string;
+    basename: string;
```
I think we can ensure this is always available from `ServerBuild` through `window.__remixContext` since we'll default it to `/`

brophdawg11 replied:
Scratch that - made it optional again in [d20b1b5](https://github.com/remix-run/remix/pull/8145/commits/d20b1b5149afe32aa9f9e4c1077a6bf3f7aee335) since non-vite builds won't have it

brophdawg11 commented on packages/remix-server-runtime/server.ts:
```diff
@@ -119,12 +118,12 @@ export const createRequestHandler: CreateRequestHandlerFunction = (
 
       response = await handleDataRequestRR(
         serverMode,
+        _build,
```
Pass the full `build` through - this aligns slightly closer to the signature for `handleDocumentRequest`

brophdawg11 commented:
Awesome work @hi-ogawa!  I did some testing and made a few small modifications - this was working well for me.  I'm going to touch base with the team this week about getting this merged 👍

hi-ogawa commented:
Hi @brophdawg11, thanks for the update on this PR! I haven't looked at this for a while, so I was afraid of merge conflict. I appreciate you've taken care of that.

As you pointed out in https://github.com/remix-run/remix/pull/8145#discussion_r1462269626, I also worry that changing `req.url -> originalUrl` in `packages/remix-express/server.ts` can cause a breakage if users are relying on custom middleware structure. It would be nice if that change is not necessary inside `createRequestHandler` but can be dealt with in user's custom server code, but I haven't figured out how to do that. I would appreciate if you can share how you are testing custom server with basename currently.

brophdawg11 commented:
@hi-ogawa Do you know of any reason we couldn't use the Vite [`base`](https://vitejs.dev/config/shared-options.html#base) config and avoid adding a config of our own (we've decided that this is OK being a vite-only feature, so no need to worry about a config flag that would wok with the old compiler).  

My last few commits update the plugin and E2E tests to use this approach and it looks good from my local testing as well.

markdalgleish commented on integration/vite-basename-dev-test.ts:
```diff
@@ -0,0 +1,140 @@
+  test.beforeAll(async () => {
+    port = await getPort();
+    cwd = await createProject({
+      "vite.config.js": await VITE_CONFIG({
+        port,
+        viteOptions: '{ base: "/mybase/" }',
+        pluginOptions: '{ publicPath: "/mybase/" }',
+      }),
```
I refactored this test to use the `VITE_CONFIG` file helper. To make this work I've also added a `viteOptions` to it.

markdalgleish commented:
@hi-ogawa Thanks for the PR! If it's not too much trouble, would you be able to add tests for custom servers (to test `@remix-run/express`) and production builds? Just want to make sure this is working across all major scenarios.

hi-ogawa commented:
> Do you know of any reason we couldn't use the Vite [base](https://vitejs.dev/config/shared-options.html#base) config and avoid adding a config of our own?

As far as I understand, Vite's `base` config is about where to serve static assets, so the concept is close to remix's [`publicPath`]( https://remix.run/docs/en/main/file-conventions/remix-config#publicpath) (e.g. it can be external domain like suggested in both documentation) and I think users should able to use `base/publicPath` independently from React routers' [`basename`](https://reactrouter.com/en/main/router-components/browser-router#basename).

hi-ogawa commented:
@markdalgleish Thanks for review! I will take a look at tests and expand further cases.

hi-ogawa commented:
I added basic test cases to cover all four patterns (vite / express) x (dev / build). I also adjusted a few test helpers to be able to use similar test setup code for basename tests.

brophdawg11 commented:
Thanks for the new tests @hi-ogawa!  I merged `dev` in and took care of the merge conflicts and did some more local testing and things are looking good.

@markdalgleish Let's sync up next week on the comments above about `base` and `publicPath`.  

I think for `npm run dev` - you _mostly_ want them to be the same thing?  I was having trouble getting them to work as different root relative paths (`/base/` versus `/base/other/`).  

With a custom express server `publicPath` is still useful to put your assets on a CDN and I was able to get it working (build/start) with `base: "/base/"` and `publicPath: "http://localhost:3001/other"`.

I'm unsure if I'm missing a capability or if we want to just recommend (or even enforce) that they're the same in dev, but let you do whatever you want in prod.

brophdawg11 commented:
Chatted with @markdalgleish and @pcattori and we realized that Vite's `base` is actually just Remix's `publicPath` (and is effectively referred to as that in the [docs](https://vitejs.dev/guide/build.html#public-base-path)), and `basename` should be a separate thing after all and not coupled to `base`.  It's totally possible in prod to use a router `basename` of `/app/` but then use a vite `base` of `http://cdn.example.com/assets` - so we can't couple them together.

I updated the plugin such that:
* Vite `base` sets `ServerBuild.publicPath`
* `remix({ basename })` passes through to the React Router [`basename`](https://reactrouter.com/en/main/routers/create-browser-router#basename) config

It's slightly confusing to see `base` and `basename` so close together in the config since they do two different things.  We're stuck with `base` from the vite side.  We _could_ change the remix plugin config value - but I hesitate to stray from the `basename` wording that so many React Router users already know.  

1. We tossed around potentially using `routerBasename` to differentiate it a bit more
2. Pedro also suggested just `base` since then there's a difference from the bundler `base` (assets) and the Remix `base` (routes) - but that could be ambiguous and tricky to document

kiliman commented:
I think using `basename` in the Vite config is fine. This is not a common usage, and those who need it will know what it means. And the fact that it maps to RR's `basename` makes it quite clear.

brophdawg11 commented on packages/remix-dev/vite/plugin.ts:
```diff
@@ -1131,7 +1150,7 @@ export const remixVitePlugin: RemixVitePlugin = (remixUserConfig = {}) => {
                   nodeRes
                 ) => {
-                  let req = fromNodeRequest(nodeReq);
+                  let req = fromNodeRequest(nodeReq, build.publicPath !== "/");
```
use the `originalUrl` when `publicPath` is not `/` indicating that a `base` was set and vite will alter `req.url`

brophdawg11 commented on packages/remix-express/server.ts:
```diff
@@ -49,7 +49,8 @@ export function createRequestHandler({
     try {
-      let request = createRemixRequest(req, res);
+      let { publicPath } = typeof build === "function" ? await build() : build;
+      let request = createRemixRequest(req, res, publicPath !== "/");
```
@markdalgleish I'm less confident in this now since it's using `publicPath` which may or may not have come from `vite` - so it could potentially introduce a breaking change if existing Remix apps with a defined `publicPath` are using (and relying on) middlewares that mess with `req.url`.  We may need either a future flag, or ideally a way to detect that this is a vite build.  Maybe we could put a flag on `ServerBuild`?  Or is there something we can look for in there already that tells us with certainty that this was a vite build?

markdalgleish replied:
We could add a `compiler` export to `ServerBuild` which is `'vite' | 'esbuild'`?

markdalgleish commented on packages/remix-dev/vite/node-adapter.ts:
```diff
@@ -34,14 +31,27 @@ function fromNodeHeaders(nodeHeaders: IncomingHttpHeaders): Headers {
+  let path = useOriginalUrl ? nodeReq.originalUrl : nodeReq.url;
+  invariant(
+    path,
+    "Expected `nodeReq.originalUrl`/`rnodeReqeq.url` to be defined"
```
```suggestion
    "Expected `nodeReq.originalUrl`/`nodeReq.url` to be defined"
```

markdalgleish commented on packages/remix-dev/vite/plugin.ts:
```diff
@@ -67,16 +66,6 @@ const CLIENT_ROUTE_EXPORTS = [
 
 const CLIENT_ROUTE_QUERY_STRING = "?client-route";
 
-// We need to provide different JSDoc comments in some cases due to differences
-// between the Remix config and the Vite plugin.
-type RemixEsbuildUserConfigJsdocOverrides = {
-  /**
-   * The URL prefix of the browser build with a trailing slash. Defaults to
-   * `"/"`. This is the path the browser will use to find assets.
-   */
-  publicPath?: SupportedRemixEsbuildUserConfig["publicPath"];
-};
-
```
Oh, it's nice to see this is gone!

dahei07 commented:
cound basename option be configured differently between client and server-side? because sometimes the reverse proxy server has path-rewrite option.

hi-ogawa commented:
> cound basename option be configured differently between client and server-side? because sometimes the reverse proxy server has path-rewrite option.

Remix/React-router needs to manage a whole user-visible URL in the same way both on client and server, so I don't think your reverse proxy scenario is supported at the remix level.

Though I'm not too familiar with reverse proxy or path-rewrite, I would guess removing such path-rewrite option and letting remix app handle full url pathname would be a simpler way to make it work.

Even if some reverse proxy rewrite is unavoidable, I think it's still possible to compensate for that at your (express) custom server level, something like:

```ts
const app = express();

// do whatever necessary so that remix handler will see `req.url` with `basename`
app.use((req, _res, next) => {
  req.url = "/mybase" + req.url; 
  next();
})

// normal remix request handler
app.all("*", createRequestHandler({ ... }))
```

brophdawg11 commented on .changeset/vite-deprecate-public-path.md:
```diff
@@ -0,0 +1,7 @@
+
+Vite: Remove the ability to pass `publicPath` as an option to the Remix vite plugin
+  - ⚠️ **This is a breaking change for projects using the unstable Vite plugin with a `publicPath`**
+  - This is already handled in Vite via the [`base`](https://vitejs.dev/guide/build.html#public-base-path) so we now set the Remix `publicPath` from the Vite `base` config
```
```suggestion
  - This is already handled in Vite via the [`base`](https://vitejs.dev/guide/build.html#public-base-path) config so we now set the Remix `publicPath` from the Vite `base` config
```

AJK74 commented:
Thanks for this - in vite.config.ts we're using 
```ts
  base:"/shopify-setup-app/",
  plugins: [
    remix({basename: "/shopify-setup-app/",...})
  ]
```
and works nicely. 

Just need some docs now! 😊
